connection.go

```diff
@@ -1437,39 +1436,145 @@ func (c *Conn) handleFrames(
 	}
 	handshakeWasComplete := c.handshakeComplete
 	var handleErr error
+	var frame wire.Frame
+	var err error
+	var l int
+	var frameType wire.FrameType
+
 	for len(data) > 0 {
-		l, frame, err := c.frameParser.ParseNext(data, encLevel, c.version)
+		frameType, l, err = c.frameParser.ParseType(data, encLevel)
 		if err != nil {
+			if err == io.EOF {
+				break
+			}
 			return false, false, nil, err
 		}
 		data = data[l:]
-		if frame == nil {
-			break
-		}
-		if ackhandler.IsFrameAckEliciting(frame) {
+
+		if ackhandler.IsFrameTypeAckEliciting(frameType) {
 			isAckEliciting = true
 		}
-		if !wire.IsProbingFrame(frame) {
+		if !wire.IsProbingFrameType(frameType) {
 			isNonProbing = true
 		}
-		if log != nil {
-			frames = append(frames, toLoggingFrame(frame))
-		}
-		// An error occurred handling a previous frame.
-		// Don't handle the current frame.
-		if handleErr != nil {
+
+		// Fast path: We're inlining common cases, to avoid using interfaces
+		if frameType.IsStreamFrameType() {
+			streamFrame, l, err := wire.ParseStreamFrame(data, frameType, c.version)
+			if err != nil {
+				return false, false, nil, err
+			}
+
+			data = data[l:]
+			if streamFrame == nil {
+				break
+			}
```

Comment:
Why would that happen? `ParseStreamFrame` should always return a non-nil frame if `err == nil`.